### PR TITLE
feat: enhanced completion

### PR DESCRIPTION
**Pull request by alexgarel**
Improvements on taxonomies completion API:

* add the option of a preprocessor for taxonomies
* add a preprocessor for Open Food Facts to handle:
  * brands taxonomy
  * add main language
  * add xx entries to all languages
* fix to have names in synonyms
* add a score to completion to first match shortest entries
* enable querying multiple languages at once
* added tests
* suggest are now declared using a component inside search-bar, enabling more  parameters to be supported in the future.
* refactor web components to put interfaces in separate files
* fixes on facets to support terms with "-" and multiple terms

TODO:
- [x] modifiy search-bar web component to use langs instead of lang.

### CODE REVIEW

File: app/query.py [hunk 1]
```diff
@@ -331,28 +331,31 @@ def build_completion_query(
 
     :param q: the user autocomplete query
     :param taxonomy_names: a list of taxonomies we want to search in
-    :param lang: the language we want search in
+    :param langs: the language we want search in
```
Comment (stephanegigandet) [inline app/query.py:334]:
```suggestion
    :param langs: the languages we want search in
```

File: frontend/src/search-suggester.ts
Comment (alexgarel) [inline frontend/src/search-suggester.ts:1]: Add this component to documentation (and maybe tutorial ?)

**Comment by raphael0202** [comment]
In the search bar, when typing a single letter and deleting it, an autocomplete query with an empty `q` to the server:

http://search.localhost:8000/autocomplete?q=&langs=en&taxonomy_names=brands,categories&size=5

**Comment by raphael0202** [comment]
In `en` language, when searching `chicken`, I have the following results:

Capture d’écran du 2024-11-06 10-34-32

I was expecting `Chickens` as first result instead of `chicken meat` (both terms are linked to the same taxonomy entry `en:chickens`).

Also, the search results are not normalized: some are capitalized, some are not.

**Comment by raphael0202** [comment]
Clicking on a brand suggested by auto-completion does not seem to work. For example, with `Casino`, we have 0 results when clicking on autosuggestion, while clicking on `casino` on the left brand panel returns 4.3k results.
It looks like it's due to wrong language prefix added when clicking on autosuggestion:

http://search.localhost:8000/static/off.html?off_q=&off_facetsFilters=brands%3A%22en%3Acasino%22&off_page=1

**Review by raphael0202 — CHANGES_REQUESTED** [review]
It looks good to me, I added a few remarks about possible bugs above!

**Comment by alexgarel** [comment]
> Clicking on a brand suggested by auto-completion does not seem to work. For example, with Casino, we have 0 results when clicking on autosuggestion, while clicking on casino on the left brand panel returns 4.3k results.
> It looks like it's due to wrong language prefix added when clicking on autosuggestion

The prefix is not wrong in regard to the taxonomy… but the brands_tags field is different from the rest of the _tags field because it has no language prefix…

I will remove the language prefix for now.

**Comment by alexgarel** [comment]
> In en language, when searching chicken, I have the following results:
> I was expecting Chickens as first result instead of chicken meat (both terms are linked to the same taxonomy entry en:chickens).

> Also, the search results are not normalized: some are capitalized, some are not.

I tried to change entries order in the array, but it does not change… and don't know how to change that in Elasticsearch…

So I added the entry name in the label, this way it's more easy to understand.

**Comment by teolemon** [comment]
We could get a category image using wikidata: https://wikidata.org/w/rest.php/wikibase/v1/entities/items/Q20034?_fields=type,labels,descriptions,aliases,statements,sitelinks (image is P18)

**Review by raphael0202 — APPROVED** [review]

**Pull request merged**